### DISCUSSION TITLE
Read-only file system (Mac)

### DISCUSSION THREAD

**Discussion by MichielMoe** [lima-vm/lima#1453] [Q&A]
### Description

As I'm trying to use apptainer (Mac) to install SmartMocap, it keeps on giving me the error that the SmartMocap folder is a read-only file system:

Unable to create build: failed to create build parent dir: makedir .../SmartMocap/build-temp-2943408036: read-only file system

I've tried various amounts of efforts (chmod etc.) to solve this but when I look at the permissions of the folder it says it is readable and writeable, while in lima it's still only readable.

Somebody who knows how to fix this?

**Comment by afbjorklund** [comment]
There was a bit of a mess with the apptainer environment variable, but having a read-only home directory is default in Lima.

* https://github.com/lima-vm/lima/pull/1125

With the suggested PR, it will leave the guest home directory as-is and instead add the host home directory as a bindpath...

But it only affects the `apptainer.lima` wrapper, not `lima apptainer`.

----

If you need an external directory mounted from the host into the container (read-write), I think you need to add it to **both**:

1) Add it to the lima mounts, with read-write - similar to the default `/tmp/lima` mount.

2) Add it to the container invocation, by using the -B flag or APPTAINER_BINDPATH env.

There are some performance implications, so things like temp files are better off using the TMPDIR in the VM itself (/tmp).

Or if it is a persistant directory like a build directory, some other location on the disk image such as /var/tmp (or whatever).

Do you have a link or some logs, on what you are trying to accomplish ?

**Reply by MichielMoe** [reply]
I'm just trying to follow the instructions of https://github.com/robot-perception-group/SmartMocap/blob/master/README.md

**Reply by afbjorklund** [reply]
Probably will need some extra features in Lima, in order to get GPU (`--nv`) device support in the VM.

**Reply by afbjorklund** [reply]
But you should be able to clone and build, from the Fedora VM. You might have to install `git` first, though.

**Reply by afbjorklund** [reply]
The usual workaround is going from the host home (read-only), to the guest home (read-write). For instance, with `cd`.

```console
[anders@lima-apptainer anders]$ touch foo
touch: cannot touch 'foo': Read-only file system
[anders@lima-apptainer anders]$ pwd
/home/anders
[anders@lima-apptainer anders]$ cd
[anders@lima-apptainer ~]$ touch foo
[anders@lima-apptainer ~]$ pwd
/home/anders.linux
```

It also has much better performance (for building), since the filesystem is not mounted from a remote location (the host).

**Reply by MichielMoe** [reply]
I've worked around the problem of the read-only file system error, the next error I get is when building the apptainer.def : 
`FATAL:   While performing build: copy Failed: chown /tmp/lima/SmartMocap/smartmocap_apptainer: permission denied`
Which I don't understand as its a folder which the apptainer created it self?

**Reply by MichielMoe** [reply]
```
+ mkdir poetry
+ curl -sSL https://install.python-poetry.org
+ POETRY_HOME=/poetry python3 -
Retrieving Poetry metadata
```

# Welcome to Poetry!

This will download and install the latest version of Poetry,
a dependency and package manager for Python.

It will add the `poetry` command to Poetry's bin directory, located at:

/poetry/bin

You can uninstall at any time by executing this script with the --uninstall option,
and these changes will be reverted.

Installing Poetry (1.4.2): Done

Poetry (1.4.2) is installed now. Great!

To get started you need Poetry's bin directory (/poetry/bin) in your `PATH`
environment variable.

Add `export PATH="/poetry/bin:$PATH"` to your shell configuration file.

Alternatively, you can call Poetry explicitly with `/poetry/bin/poetry`.

You can test that everything is set up by executing:

`poetry --version`

```
+ /poetry/bin/poetry config virtualenvs.in-project true
INFO:    Adding environment to container
INFO:    Creating sandbox directory...
FATAL:   While performing build: copy Failed: chown /tmp/lima/SmartMocap/smartmocap_apptainer: permission denied
```


This was the last error that I got
